pkg/backend/httpstate/state.go

```diff
@@ -48,53 +48,61 @@ type tokenSource struct {
 	done     chan bool
 }
 
-func newTokenSource(ctx context.Context, token string, backend *cloudBackend, update client.UpdateIdentifier,
+func newTokenSource(
+	ctx context.Context,
+	initialToken string,
+	backend *cloudBackend,
+	update client.UpdateIdentifier,
 	duration time.Duration) (*tokenSource, error) {
 
-	// Perform an initial lease renewal.
-	newToken, err := backend.client.RenewUpdateLease(ctx, update, token, duration)
-	if err != nil {
-		return nil, err
+	requests, done := make(chan tokenRequest), make(chan bool)
+	ts := &tokenSource{requests: requests, done: done}
+	go ts.handleRequests(ctx, initialToken, backend.client, update, duration)
+	return ts, nil
+}
+
+func (ts *tokenSource) handleRequests(
+	ctx context.Context,
+	initialToken string,
+	client *client.Client,
+	update client.UpdateIdentifier,
+	duration time.Duration) {
+
+	// We will renew the lease after 50% of the duration has elapsed to allow time for retries.
+	renewTicker := time.NewTicker(duration / 2)
+	defer renewTicker.Stop()
+
+	state := struct {
```

Comment:
@RobbieMcKinstry how about an anon struct?

Reply:
Ahah this might be overkill for two local variables! I think you solved my issue by moving most of the work into a top-level function!